backend/fetcher.py

```diff
@@ -160,193 +159,156 @@ def fetch_spec_file(self) -> Optional[dict[str, str]]:
-        request_endpoint = self.task_request[0]
-        if not request_endpoint.startswith("tasks"):
-            # other methods of getting logs yet unknown
-            raise HTTPException(
-                detail=f"Task {self.build_or_task_id} not findable under {self.koji_pkgs_url} URL",
-                status_code=HTTPStatus.BAD_REQUEST,
-            )
-
-        if request_endpoint.endswith(".src.rpm"):
-            request_endpoint = request_endpoint.rsplit("/", 1)[0]
+    def get_task_request_url(self) -> str:
+        """
+        We need this:
+            'git+https://src.fedoraproject.org/rpms/libphonenumber.git#c88bd3...
+        This info is in self.task_request[0]
 
-        return f"{self.koji_pkgs_url}/{request_endpoint}"
+        But not every task has this though, buildArch
+        contains file-based path to SRPM, build and buildFromSCM has it
+        """
+        task_request_url = self.task_request[0]
+        if task_request_url.startswith("git+https"):
+            return task_request_url
+        parent_task = self.task_info['parent']
+        if parent_task:
+            task_request_url = self.client.getTaskInfo(parent_task, request=True)["request"][0]
+        if task_request_url.startswith("git+https"):
+            return task_request_url
+        raise HTTPException(
+            detail=(
+                f"Task {self.task_id}, parent task {parent_task} do not have a link to sources. "
+                "We can't locate the specfile."
+            ),
+            status_code=HTTPStatus.BAD_REQUEST,
+        )
 
     def _fetch_task_logs_from_task_id(self) -> list[dict[str, str]]:
         if self.task_info["arch"] != self.arch:
             raise HTTPException(
                 detail=(
-                    f"Bad arch of task {self.build_or_task_id}: "
+                    f"Bad arch of task {self.task_id}: "
                     f'expected: {self.arch} actual: {self.task_info["arch"]}'
                 ),
                 status_code=HTTPStatus.BAD_REQUEST,
             )
 
-        if self.task_info["method"] != "buildArch":
+        if self.task_info["method"] not in ("buildArch", "buildSRPMFromSCM"):
             raise HTTPException(
                 detail=(
-                    f"Task {self.build_or_task_id} method is "
+                    f"Task {self.task_id} method is "
                     f"{self.task_info['method']}. "
                     "Please select task with method buildArch."),
                 status_code=HTTPStatus.BAD_REQUEST,
             )
 
-        logs_url_dir = self._get_logs_url_from_task_id()
         logs = []
         for log_name in self.logs_to_look_for:
-            req = requests.get(f"{logs_url_dir}/{log_name}")
-            if not req.text:
+            try:
+                log_content = self.client.downloadTaskOutput(self.task_id, log_name)
+            except koji.GenericError:
+                # checkout.log not available for buildArch
```

Comment:
could this also happen for build.log and others or only for checkout.log?

Reply:
for me it only happened for checkout.log

don't know if these log names are specific to mock or koji

Reply:
There may be other failure modes, but that depends solely on how koji API treats exceptions. Looking at the source https://pagure.io/koji/blob/master/f/koji/__init__.py there are multiple more specific exceptions available. But the `GenericError` seems to be the king.

If we later encounter more issues from this API call we can always add some comments.